### ISSUE TITLE
Update CKEditor & remove deprecated webpack integration

### ISSUE THREAD

**Issue by ChristopherChudzicki** [mitodl/ocw-studio#2054]
# Description/Context
We should:
- [ ] Update CKEditor to a newer version, probably the latest
    - Studio uses v35 and community provided typescript definitions. CKEditor has first-class support for Typescript starting in v38, so we should remove the community provided type definitions. 
- [ ] Stop using [`@ckeditor/ckeditor5-dev-webpack-plugin`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-webpack-plugin), which is deprecated. 

# Plan/Design
For updating the webpack integration, the CKEditor5 guide [Integrating from source using webpack](https://ckeditor.com/docs/ckeditor5/latest/installation/advanced/alternative-setups/integrating-from-source-webpack.html) describes what to do. See also implementation in MIT Open.

One issue is that starting in v39, CKEditor includes some small branding. See [Managing the “Powered by CKEditor” logo](https://ckeditor.com/docs/ckeditor5/latest/support/licensing/managing-ckeditor-logo.html). Here's [one way](https://github.com/mitodl/mit-open/blob/ce2ecbfcff13d6bfa58fda9f94aa86e07f3aaf1d/frontends/ol-ckeditor/src/styles.scss#L3) to remove it. *Or we could leave it, since Studio is just internal anyways.*

**Comment by umar8hassan** [comment]
Current version is 35.4.0, latest is 48.3.1. Bullet 2 (Stop using [`@ckeditor/ckeditor5-dev-webpack-plugin`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-webpack-plugin)) is already done: `dev-webpack-plugin` is gone, we use `dev-translations`. MIT Open is at v38 on the same legacy webpack path we already have, last touched July 2023.

A few blockers:

1. **Markdown layer.** `turndown.ts` mutates CKEditor's shared `turndownService` singleton. That export was gone by v43, and v46 dropped turndown entirely for remark/rehype. Has to be solved at any target version. Cheap to fix, since we already depend on `turndown` and `showdown` directly and only use `markdown-gfm` for the singleton plus the `GFMDataProcessor` base class. Roughly 30 lines to own in-repo.
2. **`ckeditor5-math` is abandoned.** npm latest is 36.0.2 (Feb 2023), the `@isaul32` fork tops out at 43.2.0 (Oct 2024). Nothing supports v44+. About 1,200 lines, and we use the full widget. This is the hard blocker on reaching latest.
3. **Mechanical but wide.** No more `src/` imports in v48, v46 renamed the model classes (`Writer`, `Range`, `SchemaContext`, `DiffItem`), `LinkUI#actionsView` is now `toolbarView` so `ResourceLinkUI` needs a rewrite, and the whole webpack CKEditor block gets deleted in favour of the single `ckeditor5` package. Deleting the community typings shim closes bullet 1.


**License Requirements:**

v44 makes `config.licenseKey` required. Without it the editor throws `license-key-missing` and never starts. Fix is setting `licenseKey: 'GPL'`, which needs no account, no key and no secret. 

**Plan**

0. Own turndown and the data processor in-repo, no version change.
1. Single `ckeditor5` package at 48, covering imports, webpack, typings, `licenseKey` and the renames.
2. Math: vendor in-repo, fork under mitodl, or write a minimal replacement.
3. Cleanups and the logo call.

Alternatively we can upgrade to intermediate v43 predates the mandatory licenseKey in v44.

**Comment by pdpinch** [comment]
@ChristopherChudzicki @gumaerc can we get your input on this? In particular: 

- how bad is the ckeditor5-math situation? That's a very popular functionality 
- do you think switching to a GPL license has implications for the ocw-studio license? (more a note for myself) 
- Should we even keep using ckeditor??

**Comment by pdpinch** [comment]
@umar8hassan Can we at least upgrade to v38 in order to get the TypeScript support?

**Comment by umar8hassan** [comment]
> [@umar8hassan](https://github.com/umar8hassan) Can we at least upgrade to v38 in order to get the TypeScript support?

Yes and it'd make sense if you want to do this gradually with intermediate version stops.

Also, we might be able to upgrade till 41.4.2 in while achieving the same goals with conditioning only a few things like changing List imports to LegacyList which works till v43. 

I'm working on it and will put in a draft.

**Comment by ChristopherChudzicki** [comment]
I discussed the `ckeditor5-math` issue a bit with Claude today. Its suggestion was to just vendor (copy) source into our repo and update it. Claude pointed out that was the approach taken by [Trillium](https://github.com/TriliumNext/Trilium)
 (see https://github.com/TriliumNext/Trilium/tree/1909f217c69048b36870c6542d7e5a81c2d743af/packages/ckeditor5/src/plugins/math), who updated `ckeditor5-math` to version 48 and have added new tests for it. We could copy Trillium's, or use our own. (Trillium's licensive might be more restrictive, though they copied the very permissive CKEditor5-math license into their repo, suggesting that directory is intended to have the very permissive license).

**Comment by pdpinch** [comment]
The upgrade to v41 is on RC now. We should do some more testing on that.

**Comment by pdpinch** [comment]
Regarding @ChristopherChudzicki's suggestion, I'll take a look at the license to see if we can vendor in their fork of the Math plugin.
